src/System.Devices.Gpio/System.Devices.Gpio/RaspberryPiDriver.cs

```diff
@@ -0,0 +1,828 @@
+        private static extern int open([MarshalAs(UnmanagedType.LPStr)] string pathname, FileOpenFlags flags);
+
+        [DllImport(LibraryName, SetLastError = true)]
+        private static extern int close(int fd);
+
+        [Flags]
+        private enum MemoryMappedProtections
+        {
+            PROT_NONE = 0x0,
+            PROT_READ = 0x1,
+            PROT_WRITE = 0x2,
+            PROT_EXEC = 0x4
+        }
+
+        [Flags]
+        private enum MemoryMappedFlags
+        {
+            MAP_SHARED = 0x01,
+            MAP_PRIVATE = 0x02,
+            MAP_FIXED = 0x10
+        }
+
+        [DllImport(LibraryName, SetLastError = true)]
+        private static extern IntPtr mmap(IntPtr addr, int length, MemoryMappedProtections prot, MemoryMappedFlags flags, int fd, int offset);
+
+        [DllImport(LibraryName, SetLastError = true)]
+        private static extern int munmap(IntPtr addr, int length);
+
+        #endregion
+
+        private const string GpioMemoryFilePath = "/dev/gpiomem";
+        private const int GpioBaseOffset = 0;
+
+        private RegisterView* _registerViewPointer = null;
+
+        public RaspberryPiDriver()
+        {
+            var fileDescriptor = open(GpioMemoryFilePath, FileOpenFlags.O_RDWR | FileOpenFlags.O_SYNC);
+
+            if (fileDescriptor < 0)
+            {
+                throw new IOException($"open error number: {Marshal.GetLastWin32Error()}");
+            }
+
+            //Console.WriteLine($"file descriptor = {fileDescriptor}");
+
+            var mapPointer = mmap(IntPtr.Zero, Environment.SystemPageSize, MemoryMappedProtections.PROT_READ | MemoryMappedProtections.PROT_WRITE, MemoryMappedFlags.MAP_SHARED, fileDescriptor, GpioBaseOffset);
+
+            if (mapPointer.ToInt32() < 0)
+            {
+                throw new IOException($"mmap error number: {Marshal.GetLastWin32Error()}");
+            }
+
+            //Console.WriteLine($"mmap returned address = {mapPointer.ToInt32():X16}");
+
+            close(fileDescriptor);
+
+            _registerViewPointer = (RegisterView*)mapPointer;
+        }
+
+        public override void Dispose()
+        {
+            if (_registerViewPointer != null)
+            {
+                munmap((IntPtr)_registerViewPointer, 0);
+            }
+        }
+
+        public override void SetPinMode(int pin, GpioPinMode mode)
+        {
+            if (pin < 0 || pin >= PinCount) throw new ArgumentOutOfRangeException(nameof(pin));
+            if (!Enum.IsDefined(typeof(GpioPinMode), mode)) throw new ArgumentOutOfRangeException(nameof(mode));
+
+            switch (mode)
+            {
+                case GpioPinMode.Input:
+                case GpioPinMode.InputPullDown:
+                case GpioPinMode.InputPullUp:
+                    SetInputPullMode(pin, PinModeToGPPUD(mode));
+                    break;
+            }                
+
+            SetPinMode(pin, PinModeToGPFSEL(mode));
```

Comment:
My ignorance here: do you really want to SetPinMode after SetInputPullMode?

Reply:
Yes, because SetInputPullMode will just set the pull up/down mode and we still need to set the input/output mode using SetPinMode.